gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/keyedgrouping/KeyedGrouping.java

```diff
@@ -0,0 +1,238 @@
+  }
+
+  /**
+   * Get the internal representation of the forwarded fields for the {@link ReduceEdgeTuples} step.
+   * The forwarded fields for this step will be the grouping keys.
+   *
+   * @return A string containing the field names of all forwarded fields.
+   */
+  private String getInternalForwardedFieldsForEdgeReduce() {
+    return IntStream.of(getInternalEdgeGroupingKeys()).mapToObj(i -> "f" + i)
+      .collect(Collectors.joining(";"));
+  }
+
+  /**
+   * Get the internal representation of the forwarded fields for the {@link ReduceVertexTuples} step.
+   * The forwarded fields for this step will be the grouping keys.
+   *
+   * @return A string containing the field names of all forwarded fields.
+   */
+  private String getInternalForwardedFieldsForVertexReduce() {
+    return IntStream.of(getInternalVertexGroupingKeys()).mapToObj(i -> "f" + i)
+      .collect(Collectors.joining(";"));
+  }
+
+  /**
+   * Get the internal representation of the forwarded fields for the {@link UpdateIdField} steps.
+   * Forwarded fields will be all but the currently updated field for the left side and the new id for
+   * the right side.
+   *
+   * @param targetField The index of the updated field.
+   * @param fromSecond  Return the forwarded fields for the right side of the join if set to true (returns
+   *                    the fields for the left side otherwise).
+   * @return The string containing the forwarded fields info.
+   */
+  private String getInternalForwardedFieldsForEdgeUpdate(int targetField, boolean fromSecond) {
+    if (fromSecond) {
+      return "f1->f" + targetField;
+    } else {
+      return IntStream.range(0,
+        GroupingConstants.EDGE_TUPLE_RESERVED + edgeGroupingKeys.size() + edgeAggregateFunctions.size())
+        .filter(i -> i != targetField).mapToObj(i -> "f" + i).collect(Collectors.joining(";"));
+    }
+  }
+
+  /**
+   * Get the internal grouping keys used for grouping the edge tuples.
+   *
+   * @return The grouping keys, as tuple indices.
+   */
+  private int[] getInternalEdgeGroupingKeys() {
+    return IntStream.range(0, GroupingConstants.EDGE_TUPLE_RESERVED + edgeGroupingKeys.size())
+      .toArray();
+  }
+
+  /**
+   * Get the internal grouping keys used for grouping the vertex tuples.
+   *
+   * @return The grouping keys, as tuple indices.
+   */
+  private int[] getInternalVertexGroupingKeys() {
+    return IntStream.range(GroupingConstants.VERTEX_TUPLE_RESERVED,
+      GroupingConstants.VERTEX_TUPLE_RESERVED + vertexGroupingKeys.size()).toArray();
+  }
+
+  /**
+   * Enable or disable an optional combine step before the reduce step.
+   * Note that this currently only affects the edge reduce step.
+   * <p>
+   * The combine step is enabled by default.
+   *
+   * @param useGroupCombine {@code true}, if a combine step should be used.
+   * @return This operator.
+   */
+  public KeyedGrouping<G, V, E, LG, GC> setUseGroupCombine(boolean useGroupCombine) {
```

Comment:
This seems untested. Have you run the tests with this `false`?

Reply:
Yes, this works when this is set to false. The combine step does not do anything special, we have to benchmark the effects of turning this on or off and maybe remove this afterwards